src/shim/shim_syscall.c

```diff
@@ -0,0 +1,134 @@
+/*
+ * The Shadow Simulator
+ * See LICENSE for licensing information
+ */
+
+#include <assert.h>
+#include <errno.h>
+#include <stdarg.h>
+#include <stdbool.h>
+#include <sys/syscall.h>
+#include <sys/time.h>
+#include <time.h>
+
+#include "shim/shim.h"
+#include "shim/shim_syscall.h"
+#include "support/logger/logger.h"
+
+#define SIMTIME_NANOS_PER_SEC 1000000000l
+#define SIMTIME_MICROS_PER_NANOSEC 1000l
+
+// We store the simulation time using timespec to reduce the number of
+// conversions that we need to do while servicing syscalls.
+static struct timespec _cached_simulation_time = {0};
+
+void shim_syscall_set_simtime_nanos(uint64_t simulation_nanos) {
+    _cached_simulation_time.tv_sec = simulation_nanos / SIMTIME_NANOS_PER_SEC;
+    _cached_simulation_time.tv_nsec = simulation_nanos % SIMTIME_NANOS_PER_SEC;
+}
+
+uint64_t shim_syscall_get_simtime_nanos() {
+    return (uint64_t)(_cached_simulation_time.tv_sec * SIMTIME_NANOS_PER_SEC) +
+           _cached_simulation_time.tv_nsec;
+}
+
+bool shim_syscall_is_supported(long syscall_num) {
```

Comment:
Instead of trying to keep this in sync with which syscalls are supported in `shim_syscall`, could the caller just try calling `shim_syscall` and fallback if it returns false?

Reply:
Yep, that was one wrinkle I noticed but left in place. But I agree and I'll change it.